### PR TITLE
Add scanning for sub projects

### PR DESCRIPTION
**Pull request by SanjulaGanepola**

### CODE REVIEW

**Review by edmundreinhardt — CHANGES_REQUESTED** [review]

File: src/iproject.ts [hunk 1]
```diff
@@ -1484,6 +1484,28 @@ export class IProject {
...
+   * 
+   * @returns The URIs for all sub projects.
+   */
+  public async getSubIProjectUris(): Promise<Uri[]> {
+    const subDirectoryUris = (await workspace.fs.readDirectory(this.workspaceFolder.uri))
+      .filter((folder) => folder[1] === FileType.Directory)
+      .map((folder) => Uri.joinPath(this.workspaceFolder.uri, folder[0]))
+
+    const subIProjectUris: Uri[] = [];
+    for await (const uri of subDirectoryUris) {
```
[the remark this answers is not in the document]
Comment (SanjulaGanepola) [reply src/iproject.ts:1499]: When you say number of directories to scan, do you mean the maximum number of directories to scan at any level or the maximum number of levels to scan (defaulted at 2)?

To scan a directory I am using the VS Code `FileSystem.readDirectory` API which retrieves all directories at the given path. It doesn't look like it has any parameters at the moment to limit the number of retrieved directories.

[the remark this answers is not in the document]
Comment (SanjulaGanepola) [reply src/iproject.ts:1499]: If it is the depth we want as a setting, I can add that. I will keep the default as 2

**Comment by SanjulaGanepola** [comment]
@edmundreinhardt I have pushed changes (https://github.com/IBM/vscode-ibmi-projectexplorer/pull/476/commits/705d6c5e7c850337d328fa04064fa49b42d3399d) to include what we discussed with William:

1. `Scan For Projects` action to manually trigger a re-scan which can be used if the user manually git clones a project from the terminal:
image

2. New preference to set scan depth:
image

3. Initial scan only happens now if the workspace folder does not have an `iproj.json`

4. Updated the `scanAndAddSubIProjects` API to take in a `uri` so that it can be used by IBM i Developer.

File: src/projectManager.ts [hunk 2]
```diff
@@ -367,10 +376,51 @@ export class ProjectManager {
...
+                const iprojUri = Uri.file(path.join(uri.fsPath, 'iproj.json'));
+                await workspace.fs.stat(iprojUri);
+
+                // Project found
+                subIProjectUris.push(uri);
+            } catch (e) {
+                // Project not found
+            }
+
+            if (currentDepth != scanDepth) {
```
Comment (edmundreinhardt) [inline src/projectManager.ts:419]: If we find a project at 1 level, we should not look in its subdirectories

**Review by edmundreinhardt — CHANGES_REQUESTED** [review]

**Review by edmundreinhardt — APPROVED** [review]
Looks good, the recursive approach should work fine

**Review by william-xiang — APPROVED** [review]
Great job!

**Comment by SanjulaGanepola** [comment]
I will merge this PR later on in case we need a new Project Explorer release for v2 with fixes that exclude this change

**Comment by edmundreinhardt** [comment]
> I will merge this PR later on in case we need a new Project Explorer release for v2 with fixes that exclude this change

agreed

**Pull request merged**